products/swift-Ninja/NinjaManifest.swift

```diff
@@ -1,169 +1,154 @@
+    self.name = name
+    self.variables = variables
+  }
 }
 
-/// An individual command in a Ninja manifest.
-public struct Command: Codable, Equatable {
-    /// The name of the rule which produced this command.
-    public let rule: String
+public struct NinjaBuildStatement: Codable, Equatable {
+  public let rule: NinjaRule
+  public let command: String
+  public let description: String
+  public var allInputs: [String] {
+    return explicitInputs + implicitInputs + orderOnlyInputs
+  }
+  public let explicitInputs: [String]
+  public let implicitInputs: [String]
+  public let orderOnlyInputs: [String]
+  public let outputs: [String]
+  public let variables: [String: String]
+  public let generator: Bool
+  public let restat: Bool
+
+  public init(rule: NinjaRule, command: String,
+              description: String, explicitInputs: [String],
+              implicitInputs: [String], orderOnlyInputs: [String],
+              outputs: [String], variables: [String: String],
+              generator: Bool, restat: Bool) {
+    self.rule = rule
+    self.command = command
+    self.description = description
+    self.explicitInputs = explicitInputs
+    self.implicitInputs = implicitInputs
+    self.orderOnlyInputs = orderOnlyInputs
+    self.outputs = outputs
+    self.variables = variables
+    self.generator = generator
+    self.restat = restat
+  }
+}
 
-    /// The list of regular inputs to the command.
-    public let inputs: [String]
+public struct NinjaManifest: Codable, Equatable {
+  public let rules: [String: NinjaRule]
+  public let statements: [NinjaBuildStatement]
+  public let defaultTargets: [String]
+  public let error: String?
+
+  public init(rules: [String: NinjaRule], statements: [NinjaBuildStatement],
+              defaultTargets: [String]) {
+    self.rules = rules
+    self.statements = statements
+    self.defaultTargets = defaultTargets
+    self.error = nil
+  }
+}
 
-    /// The list of order only inputs to the command.
-    public var implicitInputs: [String] {
-         return _implicitInputs ?? []
-    }
-    private let _implicitInputs: [String]?
- 
-    /// The list of order only inputs to the command.
-    public var orderOnlyInputs: [String] {
-        return _orderOnlyInputs ?? []
+extension NinjaManifest {
+  public init(path: String, workingDirectory: String) {
+    var cManifest = llb_manifest_fs_load(path, workingDirectory)
+    defer {
+      llb_manifest_destroy(&cManifest)
     }
-    private let _orderOnlyInputs: [String]?
-
-    /// The list of outputs of the command.
-    public let outputs: [String]
 
-    /// The command itself (a string to run with "/bin/sh -c").
-    public let command: String
-
-    /// The description of the string, or empty.
-    public let description: String
-
-    /// Whether or not the command is a generator.
-    public var generator: Bool {
-        return _generator ?? false
+    if cManifest.error.length > 0 {
+      self.error = ownedString(cManifest.error)
```

Comment:
Should this part throw an error instead? And is there any concern with continuing even after an error?

Reply:
My thought was that it depends on the client and that it may still be useful to have the rest of the manifest, even if there were errors. Ie. maybe they only care about a particular build statement.

I considered adding a `throwOnError` argument, but then errors would have to be handled regardless since the signature would be `throws`. Perhaps that's not too bad since `try!` could be used, thoughts?

Reply:
> it may still be useful to have the rest of the manifest, even if there were errors. Ie. maybe they only care about a particular build statement.

Maybe have a separate creation function for this? I feel that the initializer that will be commonly used should just throw the error (that would be "Swifty") and if a client has a use case to try to get info after errors then it could use another function.

Reply:
Sounds good, thanks!